### PR TITLE
[BugFix] IRSubstitution should keep Var type consistency

### PR DESCRIPTION
**Pull request by ganler**
Very similar to https://github.com/apache/tvm/pull/11582

Can be triggered by pool2d->binary op with integer 64 in attributes (the attached test case fail on GPU before this fix). cc: @junrushao1994

### CODE REVIEW

File: src/tir/ir/stmt_functor.cc [hunk 1]
```diff
@@ -640,10 +641,44 @@ Stmt IRTransform(Stmt ir_node, const runtime::PackedFunc& f_preorder,
...
 }
 
+#define DEFINE_BIOP_EXPR_PROMOTABLE_MUTATE_(OP)                                           \
+  PrimExpr VisitExpr_(const OP##Node* op) override {                                      \
+    PrimExpr a = this->VisitExpr(op->a);                                                  \
+    PrimExpr b = this->VisitExpr(op->b);                                                  \
+    if (a.same_as(op->a) && b.same_as(op->b)) {                                           \
+      return GetRef<PrimExpr>(op);                                                        \
+    } else {                                                                              \
+      if (a.dtype().code() == b.dtype().code() && a.dtype().bits() != b.dtype().bits()) { \
```
Comment (ganler) [inline src/tir/ir/stmt_functor.cc:651]: IR substituion will eventually fail if `a.dtype() != b.dtype()` which is required by binary operators at construction time (`OP(a, b)`). By promoting the operand types automatically when their base dtype (e.g., int or float) is the same can pass the test case (which is a valid model and failed previously due to `int32` in pool2d's topi conflicts `int64` in GPU attributes.

**Comment by kparzysz-quic** [comment]
Substituting an expression of one type for a variable of a different type is an error.  We should find places where this happens and fix it there.

**Comment by kparzysz-quic** [comment]
I actually have a patch that fixes your problem without changing substitution, I can create a PR tomorrow.

**Comment by ganler** [comment]
Yeah, I think @kparzysz-quic is right. We should look for the root cause. Looking forward to the PR!

**Pull request closed without merging**